### PR TITLE
cdc-sink: Convert to pgx library

### PR DESCRIPTION
**Pull request by bobvawter**
This change converts cdc-sink to use the pgx API directly. The tests are green against 21.1.6.

### CODE REVIEW

File: main_test.go [hunk 1]
```diff
@@ -1013,57 +1161,134 @@ func TestMultipleFeeds(t *testing.T) {
...
+			return errors.Wrap(err, "querying from")
+		}
+		if toCount == fromCount {
+			break
+		}
+	}
+	return nil
+}
+
+func loopUntilSync(
```
Comment (BramGruneir) [inline main_test.go:1271]: Please add a comment here. Just to describe what sync means.

File: main_test.go [hunk 2]
```diff
@@ -1013,57 +1161,134 @@ func TestMultipleFeeds(t *testing.T) {
...
+							break
+						}
 					}
 				}
 			}
 		})
 	}
 }
+
+func loopUntilMaxB(
```
Comment (BramGruneir) [inline main_test.go:1246]: commend here please

File: main_test.go [hunk 3]
```diff
@@ -578,8 +706,8 @@ func TestTypes(t *testing.T) {
 		{`decimal_null`, `DECIMAL`, ``, false},
 		{`float`, `FLOAT`, `1.2345`, true},
 		{`float_null`, `FLOAT`, ``, false},
-		{`geography`, `GEOGRAPHY`, `0101000020E6100000000000000000F03F0000000000000040`, false},
-		{`geometry`, `GEOMETRY`, `010100000075029A081B9A5DC0F085C954C1F84040`, false},
+		//		{`geography`, `GEOGRAPHY`, `0101000020E6100000000000000000F03F0000000000000040`, false},
```
Comment (BramGruneir) [inline main_test.go:709]: Geospatials don't work?

File: main_test.go [hunk 4]
```diff
@@ -709,104 +830,112 @@ func TestTypes(t *testing.T) {
...
 			// Wait until the out table has a row.
-			for tableOut.getTableRowCount(t) != 1 {
-				// add a stopper here from a wrapper around the handler.
-
+			for {
+				count, err := tableOut.getTableRowCount(ctx)
+				if !a.NoError(err) {
+					return
+				}
+				if count > 0 {
```
Comment (BramGruneir) [inline main_test.go:915]: `> 0` is different than `!= 1`  Could it have more than one, if it doesn't, isn't that an error?

File: main_test.go [hunk 5]
```diff
@@ -1013,57 +1161,134 @@ func TestMultipleFeeds(t *testing.T) {
...
 					}
 				}
 			}
 
 			// Delete all rows in the table.
-			// Update all rows in the source table.
 			for _, feedTables := range sourceTablesByFeed {
 				for _, table := range feedTables {
-					table.deleteAll(t)
+					a.NoErrorf(table.deleteAll(ctx), "deleting %s", table)
```
Comment (BramGruneir) [inline main_test.go:1224]: Doesn't this need to return on error?

File: sink.go [hunk 6]
```diff
@@ -142,47 +144,15 @@ func (s *Sink) deleteRows(tx *sql.Tx, lines []Line) error {
...
 		// Upsert the line
-		if _, err := tx.Exec(statement.String(), keys...); err != nil {
+		if _, err := tx.Exec(ctx, statement.String(), keys...); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-// cleanValue will check the type of the value being upserted to ensure it
```
Comment (BramGruneir) [inline sink.go:152]: So happy to see this hack removed!

File: sink_table.go [hunk 7]
```diff
@@ -159,44 +160,30 @@ func parseLine(rawBytes []byte) (Line, error) {
...
-func CreateSinkTable(db *sql.DB, sinkTableFullName string) error {
-	return Execute(db, fmt.Sprintf(sinkTableSchema, sinkTableFullName))
+func CreateSinkTable(ctx context.Context, db *pgxpool.Pool, sinkTableFullName string) error {
+	return Execute(ctx, db, fmt.Sprintf(sinkTableSchema, sinkTableFullName))
 }
 
-// WriteToSinkTable upserts all lines to the sink table. Never submit more than
-// 10,000 lines to this function at a time.
-func WriteToSinkTable(db *sql.DB, sinkTableFullName string, lines []Line) error {
+// WriteToSinkTable upserts all lines to the sink table via the COPY protocol.
```
Comment (BramGruneir) [inline sink_table.go:167]: Much nicer

File: sink_table.go [hunk 8]
```diff
@@ -159,44 +160,30 @@ func parseLine(rawBytes []byte) (Line, error) {
...
-			if _, err := fmt.Fprintf(&statement, ",($%d,$%d,$%d,$%d)", j+1, j+2, j+3, j+4); err != nil {
-				return err
-			}
-		}
-	}
-
-	return Execute(db, statement.String(), values...)
+	parts := strings.Split(strings.ToLower(sinkTableFullName), ".")
+	_, err := db.CopyFrom(ctx, parts,
+		[]string{"nanos", "logical", "key", "after"},
```
Comment (BramGruneir) [inline sink_table.go:174]: Worth making this const?

File: sinks.go [hunk 9]
```diff
@@ -123,9 +121,15 @@ func (s *Sinks) HandleResolvedRequest(
 		}
 
 		// Start the transation
-		if err := crdb.ExecuteTx(context.Background(), db, nil, func(tx *sql.Tx) error {
+		if err := Retry(ctx, func(ctx context.Context) error {
```
Comment (BramGruneir) [inline sinks.go:124]: Does retry handle the crdb specific retires?  Is 'github.com/cockroachdb/cockroach-go/crdb' no longer needed?

File: util.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2020 The Cockroach Authors.
```
Comment (BramGruneir) [inline util.go:1]: Should we import this instead?  Or is it not compatible with pgx?

At least mention the providence.

**Review by BramGruneir — CHANGES_REQUESTED** [review]
Looks good overall, just a few minor questions.

File: main_test.go (hunk 2 above)
Comment (bobvawter) [reply main_test.go:1246]: Done.

File: main_test.go (hunk 3 above)
Comment (bobvawter) [reply main_test.go:709]: Added a TODO; I can't figure out yet how to get them to re-encode from the server properly.

File: main_test.go (hunk 4 above)
Comment (bobvawter) [reply main_test.go:915]: Fixed.

File: sink_table.go (hunk 8 above)
Comment (bobvawter) [reply sink_table.go:174]: Extracted to global variable.

File: util.go (hunk 10 above)
Comment (bobvawter) [reply util.go:1]: The cockroach pgx support library doesn't allow for implicit transactions. The specific codes are

```
			case "40001": // Serialization Failure
			case "40003": // Statement Completion Unknown
			case "08003": // Connection Does Not Exist
			case "08006": // Connection Failure
```

The provenance is mentioned at the top of the file, it's from Cacheroach.

File: main.go [hunk 11]
```diff
@@ -168,28 +176,43 @@ func main() {
...
-	handler := createHandler(db, sinks)
-	http.HandleFunc("/", handler)
-	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", *port), nil))
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", *port))
+	if err != nil {
+		log.Fatalf("could not open listener: %v", err)
+	}
+	log.Printf("listening on %s", l.Addr())
+
+	handler := http.Handler(http.HandlerFunc(createHandler(db, sinks)))
```
[the remark this answers is not in the document]
Comment (bobvawter) [reply main.go:206]: Fixed.

File: main_test.go (hunk 1 above)
Comment (bobvawter) [reply main_test.go:1271]: Done.

**Pull request merged**